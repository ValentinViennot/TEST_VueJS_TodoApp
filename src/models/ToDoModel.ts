export default interface ToDoModel {
  text: string;
  done: boolean;
  id: number;
}
