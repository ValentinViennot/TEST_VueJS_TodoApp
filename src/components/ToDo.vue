<template>
  <div class="todoContainer" @click="toggleDone">
      <div v-if="todo.done" class="todoIsDone true">
          [DONE]
      </div>
      <div v-else class="todoIsDone false">
          [TODO]
      </div>
      <div class="todoText">
      <p>{{ todo.text }}</p>
  </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';
import ToDoModel from '../models/ToDoModel';

@Component
export default class ToDo extends Vue {
  @Prop()
  private todo!: ToDoModel;

  private toggleDone() {
    this.todo.done = !this.todo.done;
  }
}
</script>

<style scoped>
.todoContainer {
  display: block;
  background-color: white;
  cursor: pointer;
  opacity: 0.5;
}
.todoContainer:hover {
  opacity: 1;
}

.todoContainer:nth-child(2n) {
  background-color: whitesmoke;
}

.todoContainer > * {
  display: inline-block;
  margin: 5px;
}

.todoIsDone {
  font-size: 0.5em;
  font-weight: bold;
}

.todoIsDone.true {
  color: green;
}
.todoIsDone.false {
  color: red;
}
</style>
