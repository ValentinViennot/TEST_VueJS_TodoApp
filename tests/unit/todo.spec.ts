import { shallowMount, Wrapper } from '@vue/test-utils';
import ToDo from '../../src/components/ToDo.vue';
import ToDoModel from '../../src/models/ToDoModel';

describe('ToDo.vue component', () => {
  let dummyTodo: ToDoModel;
  let wrapper: Wrapper<ToDo>;
  beforeAll(() => {
    dummyTodo = {
      text: 'test',
      done: true,
      id: 0
    };
    wrapper = shallowMount(ToDo, {
      propsData: { todo: dummyTodo }
    });
  });

  it('should render without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('when it is marked as done', () => {
    beforeAll(() => {
      dummyTodo = {
        text: 'test',
        done: true,
        id: 0
      };
      wrapper = shallowMount(ToDo, {
        propsData: { todo: dummyTodo }
      });
    });

    it('should contain [DONE] element', () => {
      expect(wrapper.findAll('.todoIsDone.true')).toHaveLength(1);
    });
    it('should not contain [TODO] element', () => {
      expect(wrapper.findAll('.todoIsDone.false')).toHaveLength(0);
    });

    describe('when component is clicked', () => {
      beforeAll(() => {
        wrapper.trigger('click');
      });

      it('should now be not done', () => {
        expect(dummyTodo.done).toBe(false);
      });
    });
  });

  describe('when it is marked as not done', () => {
    beforeAll(() => {
      dummyTodo = {
        text: 'test',
        done: false,
        id: 0
      };
      wrapper = shallowMount(ToDo, {
        propsData: { todo: dummyTodo }
      });
    });

    it('should not contain [DONE] element', () => {
      expect(wrapper.findAll('.todoIsDone.true')).toHaveLength(0);
    });
    it('should contain [TODO] element', () => {
      expect(wrapper.findAll('.todoIsDone.false')).toHaveLength(1);
    });

    describe('when component is clicked', () => {
      beforeAll(() => {
        wrapper.trigger('click');
      });

      it('should now be done', () => {
        expect(dummyTodo.done).toBe(true);
      });
    });
  });
});
