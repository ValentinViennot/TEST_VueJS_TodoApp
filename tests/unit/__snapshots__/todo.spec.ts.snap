// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDo.vue component should render without crashing 1`] = `
<div class="todoContainer">
  <div class="todoIsDone true">
    [DONE]
  </div>
  <div class="todoText">
    <p>test</p>
  </div>
</div>
`;
